File: application/backend-credit-card-enrollment/backend-c#/Common/SessionAuth/SessionRepository.cs
using MongoDB.Bson;
using MongoDB.Driver;
using CreditCardEnrollment.Common.Projection;
using System;

namespace CreditCardEnrollment.Common.SessionAuth;

public class SessionRepository {
    private readonly MongoTransactionalProjectionOperator _mongoOperator;

    public SessionRepository(MongoTransactionalProjectionOperator mongoOperator) {
        _mongoOperator = mongoOperator;
    }

    public string? AuthenticatedUserIdFromSessionToken(string sessionToken, int sessionExpirationSeconds) {
        var session = _mongoOperator
            .Find<BsonDocument>(
                "AuthenticationForAllContexts_Session_Session", 
                doc => doc["sessionToken"] == sessionToken
            ).FirstOrDefault();

        if (session == null) return null;
        
        if (session.GetValue("signedOut").AsBoolean) return null;

        var tokenLastRefreshedStr = session.GetValue("tokenLastRefreshedAt").AsString;
        var tokenLastRefreshed = DateTimeOffset.Parse(tokenLastRefreshedStr).UtcDateTime;

        if (tokenLastRefreshed < DateTime.UtcNow.AddSeconds(-sessionExpirationSeconds)) return null;

        return session.GetValue("userId").AsString;
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/common/sessionAuth/SessionRepository.ts
import { Document, Filter } from 'mongodb';
import {MongoTransactionalProjectionOperator} from "../projection/MongoTransactionalProjectionOperator";
import {inject, injectable} from "tsyringe";

@injectable()
export class SessionRepository {
    private readonly mongoOperator: MongoTransactionalProjectionOperator;

    constructor(
        @inject(MongoTransactionalProjectionOperator) mongoOperator: MongoTransactionalProjectionOperator
    ) {
        this.mongoOperator = mongoOperator;
    }

    async authenticatedUserIdFromSessionToken(
        sessionToken: string,
        sessionExpirationSeconds: number
    ): Promise<string | null> {
        const sessionCollectionName = 'AuthenticationForAllContexts_Session_Session';

        try {
            const sessions = await this.mongoOperator.find<Document>(
                sessionCollectionName,
                { sessionToken } as Filter<Document>
            );

            if (sessions.length === 0) return null;

            const session = sessions[0];
            if (session.signedOut) return null;

            const tokenLastRefreshedStr = session.tokenLastRefreshedAt as string;
            const tokenLastRefreshed = new Date(tokenLastRefreshedStr).getTime();

            if (
                tokenLastRefreshed <
                Date.now() - sessionExpirationSeconds * 1000 // Convert seconds to milliseconds
            ) {
                return null;
            }

            return session.userId as string;
        } catch (error) {
            throw new Error(
                `Error fetching user ID from session token: ${(error as Error).message}`
            );
        }
    }
}
